### DISCUSSION TITLE
Accessing I2s pins for input with Audiokit

### DISCUSSION THREAD

**Discussion by TangerineDreamer** [pschatzmann/arduino-audio-tools#1458] [Q&A] [answered]
Hi,
i want to try this possibility,'cos those pins are available on the audiokit ,to wire a teensy output, because i was disapointed to do it with 
a simple esp32 , the sound looks half speed and the wifi radio with audiokit seems to be more powerful.

i found these audiokit pinout on the net:

![Audiokit pinout](https://github.com/pschatzmann/arduino-audio-tools/assets/76820158/b05011f0-8a2e-4f17-a32b-bdf609bcd950)

So i use the pins 15,14,22 to input the teensy,my whole sketch:

```
#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"

AudioBoardStream kit(AudioKitEs8388V1); // Access I2S as stream
StreamCopy copier(kit, kit); // copy kit to kit

// Arduino Setup
void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Warning);
    
    auto cfg = kit.defaultConfig(RXTX_MODE);
    cfg.sd_active = false;
    //cfg.input_device = ADC_INPUT_LINE2;
    cfg.is_master = false;
    cfg.use_apll = false;
    cfg.pin_ws = 14; 
   cfg.pin_bck = 15;
   cfg.pin_data = 22;
    kit.begin(cfg);
}

// Arduino loop - copy data
void loop() {
    copier.copy();
}
```
Can this be realisable?

**Comment by pschatzmann** [comment, marked as answer]
I don't know: I have never tried. But taking into consideration that many people are struggling to get SPI to work with this board, it might be difficult.

Be careful to double check for pin conflicts..

**Comment by TangerineDreamer** [comment]
Yes it seems to walk on eggs
I was refering to the esp32 I2s,but i don't know how to place ws,bck,sck,data:
MTDO=?
MTCK=?
MTDI=?
MTMS=?

Don't you know?

**Comment by pschatzmann** [comment]
You will need to figure out what these pins are ? 
I would suggest that you define another I2SStream on port 1 so that you can continue to use the AudioKit I2S.
If your solution needs the master clock, it will not work since this is on pin 0 and this is not available via a header.

**Comment by TangerineDreamer** [comment]
i don't need master clock,it's better working without,when i wire  teeny as input,
but i'm a little bit confused about i2s pins,are they dedicated at the beginning or can define
wich of them can be ws,bck,data?

**Reply by pschatzmann** [reply]
Sure on the ESP32 you are flexible and you can define the i2s pins in your sketch like you have done in your first post.
I am wondering if you get a reliable connection, because connecting 2 ESP32 via I2S is not working properly.

**Comment by TangerineDreamer** [comment]
You're right,if i understand you say wiring 2 esp32 via i2s seems to be difficult,you see,my original searching was to
rely a teensy to a esp32,to hear on bluetooth the audio.
I tried with a teensy with  a wave generator wired to a esp32,surprised i was when i see in the arduino plotter a nice waveform,
but the sound was like a half speed or like a 16bits into a 32 bits or the inverse.

**Reply by pschatzmann** [reply]
That should not be: double check the sending and receiving format. 
The issue with ESP32 i2s to i2s is that you get some signal but it seems that samples get lost here and then...

**Comment by TangerineDreamer** [comment]
When  the esp32 shows generate a waveform -> i2s -> pcm 1801 the sound is not very good but the waveform looks pure
When  the esp32 shows the generate waveform in the serial plotter while sending to i2s pcm 1801,the sound is perturbed,maybe because
of two instances in the loop.

So,i always  hear from the bluetooth receiver,so the lineup is:
TEENSY(generator) -> first esp32:  (i2s input)->a2dp(source) --> second esp32 : a2dp(sink)->(i2s output)
Do you have a way to get infos from the signal,i know teensy gives 44100,16bits,2 channels

**Reply by TangerineDreamer** [reply]
I precise,i have forgotten to say that my receiver is an audiokit,so:
TEENSY(generator) -> first esp32: (i2s input)->a2dp(source) --> Audiokit : a2dp(sink)->(i2s output) speakers

**Comment by pschatzmann** [comment]
That's also what's needed by A2DP and what is used if you do not change the default configuration for I2S in the AudioTools.
I would suggest to look into other protocols to transmit the data e.g. Serial or SPI. 
In order to reduce the amount of transmitted data you can also reduce the sample rate on the teensy and resample on the ESP32 or used a compressing codec: e.g. ADPCM

ps. don't forget to set the log level to Warning on the ESP32 and avoid any Serial output in the loop. If my calculation is right you need to be able to reliably transmit at least 1.6 Mbps using 44100,2,16 and this is quite a challenge...
